compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs

```diff
@@ -2401,6 +2403,71 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
             suggested = true;
         }
+
+        if suggested {
+            return true;
+        }
+
+        let trait_def_id = trait_pred.def_id();
+        let trait_name = self.tcx.item_name(trait_def_id);
+        let trait_krate_name = self.tcx.crate_name(trait_def_id.krate);
+
+        // If there are multiple different versions of a crate in the dependency graph, there is already
+        // a suggestion designed for this purpose in the rustc_hir_typeck compiler crate
+        if self
+            .tcx
+            .all_traits_including_private()
+            .find(|def_id| {
+                def_id.krate != trait_def_id.krate
+                    && self.tcx.crate_name(def_id.krate) == trait_krate_name
+                    && self.tcx.item_name(def_id) == trait_name
+            })
+            .is_some()
+        {
+            return false;
+        }
+
+        let trait_has_same_params = |other_trait_def_id: DefId| -> bool {
+            let trait_generics = self.tcx.generics_of(trait_def_id);
+            let other_trait_generics = self.tcx.generics_of(other_trait_def_id);
+
+            if trait_generics.count() != other_trait_generics.count() {
+                return false;
+            }
+            trait_generics.own_params.iter().zip(other_trait_generics.own_params.iter()).all(
+                |(a, b)| {
+                    (matches!(a.kind, ty::GenericParamDefKind::Type { .. })
+                        && matches!(b.kind, ty::GenericParamDefKind::Type { .. }))
+                        || (matches!(a.kind, ty::GenericParamDefKind::Lifetime,)
+                            && matches!(b.kind, ty::GenericParamDefKind::Lifetime))
+                        || (matches!(a.kind, ty::GenericParamDefKind::Const { .. })
+                            && matches!(b.kind, ty::GenericParamDefKind::Const { .. }))
+                },
+            )
+        };
+        let trait_name = self.tcx.item_name(trait_def_id);
+        if let Some(other_trait_def_id) = self.tcx.all_traits_including_private().find(|def_id| {
+            trait_def_id != *def_id
+                && trait_name == self.tcx.item_name(def_id)
+                && trait_has_same_params(*def_id)
+                && self.predicate_must_hold_modulo_regions(&Obligation::new(
+                    self.tcx,
+                    obligation.cause.clone(),
+                    obligation.param_env,
+                    trait_pred.map_bound(|tr| ty::TraitPredicate {
+                        trait_ref: ty::TraitRef::new(self.tcx, *def_id, tr.trait_ref.args),
+                        ..tr
+                    }),
+                ))
+        }) {
+            err.note(format!(
+                "`{}` implements similarly named `{}`, but not `{}`",
+                trait_pred.self_ty(),
+                self.tcx.def_path_str(other_trait_def_id),
+                trait_pred.print_modifiers_and_trait_path()
+            ));
+            suggested = true;
+        }
```

Comment:
this suggestion is now in a function called `note_version_mismatch`. It is explicitly intended to apply to non-version mismatch errors.

Also, can you figure out why yet a separate impl of this behavior in `rustc_hir_typeck` exists?

I would personally expect that all three notes could be merged into a single:
- check whether there's another trait with the same name which is implemented by the self type
- if so, if the other trait is from another crate version -> report version mismatch
- if no, report the more general error